### ISSUE TITLE
优化 LRUCache读取性能

### ISSUE THREAD

**Issue by dugenkui03** [apache/incubator-dubbo#2610]
[LRUCache](https://github.com/apache/incubator-dubbo/blob/master/dubbo-common/src/main/java/org/apache/dubbo/common/utils/LRUCache.java)

问题
    查询可以加共享锁，更新时才需要加排他锁。因此可以使用信号量替代排他锁提高查询操作的性能。

思路：
    1. 必须使用公平锁：由AQS和Semaphore源码可知非公平锁在高并发情况下会导致更新操作难以执行；
    2. 信号量初始值为Integer.MAX_VALUE，读取操作则将状态值减1，更新操作则将状态值减Integer.MAX_VALUE，以此来实现查询操作加共享锁，更新操作加排他锁。

主要修改：
```java
//1. 排他锁ReentrantLock替换成信号来能Semaphore；

//2.size()->super.size()父类方法不需要获取许可，且仅在put类操作中被调用
    @Override
    protected boolean removeEldestEntry(java.util.Map.Entry<K, V> eldest) {
        return super.size() > maxCapacity;
    }
```

##### 参考：
- 《java并发编程实战》14.5 和 14.6 小节
- [信号量测试](https://github.com/dugenkui03/java-source-code/commit/740ac89a43896337a9b09b82775ebc1947684f1e)

**Comment by beiwei30** [comment]
@dugenkui03 looks like your implementation fails the unit tests.

**Comment by carryxyh** [comment]
@beiwei30 @dugenkui03 
I think it is a good idea.
Thx for your issue and pr. But if there are some performance tests, it will be more convincing.
:)

**Comment by beiwei30** [comment]
@dugenkui03 why close the pull request?

**Comment by beiwei30** [comment]
should we try out ReadWriteLock if semaphore has problem?

**Comment by carryxyh** [comment]
> should we try out ReadWriteLock if semaphore has problem?

Yes, we can. I considered ReadWriteLock, it seems the same effect with semaphore.
I will help him to fix ci problem.

**Comment by dugenkui03** [comment]
> > should we try out ReadWriteLock if semaphore has problem?
> 
> Yes, we can. I considered ReadWriteLock, it seems the same effect with semaphore.
> I will help him to fix ci problem.

thx,replace ReentrantLock with ReadWriteLock is easy,i finished this  noon.tough part is performance tests,because i need to consider the effects of the gc、jit especially the hit rate.details in [here](https://github.com/dugenkui03/blog/blob/master/todo/%E8%AF%BB%E5%86%99%E9%94%81.md).

failed in designing of the perfect solution，[this simple demo](https://github.com/dugenkui03/java-source-code/blob/master/src/test/java/dubbo/cache/Performance.java) can demonstrate that  ReadWriteLock is faster than ReentrantLock .i have add some functional tests in [PR](https://github.com/apache/incubator-dubbo/pull/2627).

**Comment by dugenkui03** [comment]
> should we try out ReadWriteLock if semaphore has problem?

bed PR, get() in `linkedmap` is not threadSafe,because it will update the position of visited key.thx for review.
